### ISSUE TITLE
len of typetracer array raises TypeError (cannot be interpreted as integer)

### ISSUE THREAD

**Issue by douglasdavis** [scikit-hep/awkward#1311]
### Version of Awkward Array

1.8.0rc4

### Description and code to reproduce

This looks to be related to #1216 -- the typetracer length issue causing the repr to break is now raising if I try to compute the meta of an object that has had len map partitioned across it:

```python
x = dak.from_json(...)
x.map_partitions(len)
```
will fail because `map_partitions` will try to call `len(x.meta)` to compute a new meta.

```python
In [95]: x
Out[95]: dask.awkward<getitem, npartitions=2>

In [96]: x.meta
Out[96]: <Array-typetracer type='?? * var * int64'>

In [97]: x.compute()
Out[97]: <Array [[1, 2, 3], [1, 2], ..., [3, ..., 4], [1, 2, 3]] type='40 * var * int64'>

In [98]: len(x.meta)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [98], in <module>
----> 1 len(x.meta)

File ~/.pyenv/versions/3.10.2/envs/dask-awkward/lib/python3.10/site-packages/awkward/_v2/highlevel.py:537, in Array.__len__(self)
    527 def __len__(self):
    528     """
    529     The length of this Array, only counting the outermost structure.
    530 
   (...)
    535     is `3`, not `5`.
    536     """
--> 537     return len(self._layout)

TypeError: 'UnknownLengthType' object cannot be interpreted as an integer
```

I traced the `__len__(self)` call for `x.meta` to be rely on the `length` attr of its `layout` (which is an `awkward._v2._typetracer.UnknownLengthType`):
```
In [111]: x.meta._layout.length
Out[111]: UnknownLength
```

**Comment by jpivarski** [comment]
What JSON data did `x` come from and what Awkward operations get applied to do the `x.map_partitions(len)`? I'm trying to reproduce this with Awkward-only, so that I can narrow down to find out what's putting an `UnknownLength` on the typetracer.

**Comment by douglasdavis** [comment]
`x` is created by loading [this line delimited JSON data](https://github.com/ContinuumIO/dask-awkward/blob/5a36e9155a6e508099e46834b81d3f7d1b724742/src/dask_awkward/testutils.py#L116-L136) twice. (I have a dumb [utility function](https://github.com/ContinuumIO/dask-awkward/blob/5a36e9155a6e508099e46834b81d3f7d1b724742/src/dask_awkward/testutils.py#L287) that I use to play with some made up data where the string is written to a temp file and loaded; dask_awkward's from_json is lazily calling `ak.from_iter` on the bytes of the temp file).

**Comment by jpivarski** [comment]
I still don't know how to reproduce your issue, but I did find some cases that resulted in `UnknownLength`, just due to the fact that typetracer arrays were initialized with `UnknownLength` and that propagated up.

PR #1312 changed the policy to use length information wherever possible, though it will always be the case that some typetracers won't know their lengths. (E.g. `nplike.arange(typetracer_has_no_data[-1])`.) That exposed a lot of places where we were checking lengths—previously, typetracers skipped over the test because `UnknownLength == integer` is always False, but now those checks all need to be guarded by `if nplike.known_shape` (which is False for typetracers; they don't know their shape).

It might unblock you, but you need to be prepared for partitions that don't know their lengths.

**Comment by douglasdavis** [comment]
> It might unblock you, but you need to be prepared for partitions that don't know their lengths.

Oh yes absolutely! Sorry for creating a little confusion here (I now see my wording was pretty bad), I'm not looking to try to keep or know length information, but to have the typetracer usage give back something like `UnknownScalar` when `len` is called. Here's a better description I hope:

My reason for calling `x.map_partitions(len)` was experimenting with implementing `dak.num(x, axis=0)`,

```python
def num(array, axis=..., ...):
    # in the case where we don't know the divisions we will need to create a task graph
    if axis == 0 and not array.known_divisions:
        all_lens = array.map_partitions(len)
        return dak.sum(all_lens)
    # if we know the divisions this is trivial
    elif axis == 0
        # return a lazy scalar wrapping the value of  array.divisions[-1]
````

Writing this code led me to create this issue because `array.map_partitions(f)` is calling `f(array.meta)`; here `f` is `len` which caused the `TypeError`. This was just some experimentation which led to something I thought may be a bug. 

Even more details: I went down this route after running into another typetracer issue with `ak.num` (was planning on opening another issue for that so to not spin this one out of control!)

**Comment by jpivarski** [comment]
> to have the typetracer usage give back something like `UnknownScalar` when `len` is called

Unfortunately, that's not going to work because Python has forbidden it:

```python
>>> class Something:
...     def __len__(self):
...         return ak._v2._typetracer.UnknownLength
... 
>>> len(Something())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'UnknownLengthType' object cannot be interpreted as an integer
```

and

```python
>>> class Something:
...     def __len__(self):
...         return ak._v2._typetracer.UnknownScalar(int)
... 
>>> len(Something())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'UnknownScalar' object cannot be interpreted as an integer
```

Actually, the typetracer does return `UnknownLength` when its `len` is called (if it _is_ unknown, and now that will happen less often, but it still can happen), and that's why you get the TypeError.

Throughout the Awkward codebase, we've replaced `len(something)` with `something.length`, where `length` is a property that might return `UnknownLength`, as a way of avoiding this problem.

**Comment by douglasdavis** [comment]
> Unfortunately, that's not going to work because Python has forbidden it:

🤦 ah yes huge brain-lapse for me. Had I remembered this I never would have opened this issue! Sorry for this noise
